yarn-project/simulator/src/avm/serialization/bytecode_serialization.test.ts

```diff
@@ -133,4 +134,87 @@ describe('Bytecode Serialization', () => {
     const expected = Buffer.concat(instructions.map(i => i.serialize()));
     expect(actual).toEqual(expected);
   });
+
+  it('Should throw an InvalidOpcodeError while deserializing an out-of-range opcode value', () => {
+    const decodeInvalid = () => {
+      const wrongOpcode: number = MAX_OPCODE_VALUE + 1;
+      const buf = Buffer.alloc(1);
+      buf.writeUint8(wrongOpcode);
+      decodeFromBytecode(buf);
+    };
+
+    expect(decodeInvalid).toThrow(InvalidOpcodeError);
+  });
+
+  it('Should throw an InvalidOpcodeError while deserializing an opcode value not in instruction set', () => {
+    const decodeInvalid = () => {
+      const instructionSet: InstructionSet = new Map<Opcode, any>([
+        [InstA.opcode, InstA.deserialize],
+        [InstB.opcode, InstB.deserialize],
+      ]);
+      const buf = Buffer.alloc(1);
+      buf.writeUint8(Opcode.AND_8); // Valid opcode but not in supplied instruction set.
+      decodeFromBytecode(buf, instructionSet);
+    };
+
+    expect(decodeInvalid).toThrow(InvalidOpcodeError);
+  });
+
+  it('Should throw an AvmParsingError while deserializing an incomplete instruction', () => {
+    const decodeIncomplete = (truncated: Buffer) => {
+      return () => decodeFromBytecode(truncated);
+    };
+
+    const instructions = [
+      new Call(
+        /*indirect=*/ 0x01,
+        /*gasOffset=*/ 0x1234,
+        /*addrOffset=*/ 0xa234,
+        /*argsOffset=*/ 0xb234,
+        /*argsSize=*/ 0xc234,
+        /*successOffset=*/ 0xf234,
+      ),
+    ];
+
+    const bytecode = encodeToBytecode(instructions);
+
+    for (let i = 1; i < 7; i++) {
```

Comment:
I'm fine with just one iteration but otherwise maybe use something like `bytecode.length - 1`? (so that the opcode is still there)

Reply:
Yep absolutely make sense. Your suggestion is better than a hardcoded constant.